cpp/util/read_private_key.cc

```diff
@@ -24,5 +27,68 @@ KeyError ReadPrivateKey(EVP_PKEY** pkey, const std::string& file) {
 }
 
 
+static ENGINE* engine = NULL;
+
+static void EngineFree() {
+  ENGINE_free(engine);
+}
+
+static void EngineFinish() {
+  ENGINE_finish(engine);
+}
+
+KeyError LoadEnginePrivateKey(EVP_PKEY** pkey, const std::string& engine_name,
+                              const std::string& engine_pre_cmds,
+                              const std::string& engine_post_cmds,
+                              const std::string& key_id) {
+  ENGINE_load_builtin_engines();
+  atexit(ENGINE_cleanup);
+
+  engine = ENGINE_by_id(engine_name.c_str());
+  if (!engine)
+    return NO_SUCH_ENGINE;
+  atexit(EngineFree);
+
+  // Temporary copy of the pre_cmds string that will be modified.
```

Comment:
This string manipulation is leaving me a bit unimpressed (I'd prefer to remain within the realm of ```std::string```, with things like ```find``` and ```substr```), but I'd have to read up on ```ENGINE_ctrl_cmd_string``` to know what is can or cannot do to its parameters first, maybe what I'm imagining wouldn't make sense...